contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,234 @@
+from __future__ import print_function
+from datetime import datetime
+import httplib2
+import logging
+import os
+import re
+
+# Uses Google's Client Library
+#   pip install --upgrade google-api-python-client
+from apiclient import discovery, errors
+from oauth2client import client
+from oauth2client import tools
+from oauth2client.file import Storage
+from oauth2client.tools import argparser
+
+SCOPES = 'https://www.googleapis.com/auth/calendar'
+CLIENT_SECRET_FILE = '.gcal_client_secret.json'
+APPLICATION_NAME = 'Zulip Google Calendar Bot'
+
+credentials = None
+http = None
+service = None
+
+# NOTE ABOUT GOOGLE'S OAUTH
+# Right now there are some duplicates of this function in Zulip.
+#
+# For example, in this case it asks for user's consent to read and write
+# their calendars. Different bots/integrations that require other
+# Google Account permissions, have to repeat this process and create
+# new credentials files, which potentially fills the user's computer with
+# a lot files, spread all over the disk.
+#
+# The proper apporach for this would be to create a script that wraps
+# this procedure, and stores all the credentials in the same place.
+#
+# The ideal situation will be when Zulip becomes capable of managing API
+# keys and user credentials, allowing the user to choose between which
+# bots have access to what. However, this belongs more to the long-term
+# plan in Zulip's roadmap.
+
+def get_credentials():
+    '''Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
+
+    Returns:
+        Credentials, the obtained credential.
+    '''
+    home_dir = os.path.expanduser('~')
+    credential_dir = os.path.join(home_dir, '.gcalbot-credentials')
+    if not os.path.exists(credential_dir):
+        os.makedirs(credential_dir)
+    credential_path = os.path.join(credential_dir,
+                                   'gcalbot.json')
+
+    store = Storage(credential_path)
+    cred = store.get()
+    if not cred or cred.invalid:
+        flow = client.flow_from_clientsecrets(os.path.join(home_dir, CLIENT_SECRET_FILE), SCOPES)
+        flow.user_agent = APPLICATION_NAME
+        flags = argparser.parse_args('--noauth_local_webserver --logging_level INFO'.split())
+        cred = tools.run_flow(flow, store, flags)
+    return cred
+
+# This function currently has no use.
+# That's because right now there's no way of running the bot
+# with CLI options, which would be the best fit for this case.
+# The ideal scenario would be running the bot with the -l
+# option to see all available calendars, and execute it
+# (through a different option) with the desired calendar.
+#
+# Something like:
+#   $ ./run.py lib/gcalendar.py -l
```

Comment:
See my discussion on Zulip.  The disconnect here is that you don't need `run.py` to run code in this library.  You only need `run.py` to run the bot.